File: Java/Huffman.java
import java.lang.ClassCastException;
import java.lang.ArithmeticException;

public class Huffman implements Tipo{
	private String data; 
	private int frequency; 

	public Huffman(String s, int f) throws ArithmeticException{
		if(f < 0)
			throw new ArithmeticException("Frequenza negativa");
		data = s; 
		frequency = f; 
	}

	public Huffman copia(){
		return new Huffman(data, frequency);
	}

	public boolean checkKey(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 

		if(data.equals(t.data))
			return true; 
		else
			return false; 		
	}

	public String getData(){
		return data; 
	}
	
	public int getFrequency(){
		return frequency; 
	}


	public boolean uguale(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 

		if(data.equals(t.data) && frequency == t.frequency)
			return true; 
		else
			return false; 
	}


	public boolean minore(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 
		
		if(frequency < t.frequency)
			return true; 
		else
			return false; 
	}

	public boolean maggiore(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 

		if(frequency > t.frequency)
			return true; 
		else
			return false; 
	}


	public boolean maggioreUguale(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passto non valido"); 
		if(frequency >= t.frequency)
			return true; 
		else
			return false; 
	}

	public boolean minoreUguale(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 
		if(frequency <= t.frequency)
			return true; 
		else
			return false; 
	}

	public Tipo somma(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 
		return new Huffman(data + t.data, frequency + t.frequency);
	}

	public Tipo sottrazione(Tipo t) throws ClassCastException{
		throw new ClassCastException("Operazione di sottrazione non supportata"); 
	}

	public Tipo moltiplicazione(Tipo t) throws ClassCastException{
		throw new ClassCastException("Operazione di moltiplicazione non supportata"); 
	}

	public Tipo divisione(Tipo t) throws ClassCastException{
		throw new ClassCastException("Operazione di moltiplicazione non supportata"); 
	}

	public String toString(Tipo t) throws ClassCastException{
		if(!(t instanceof Huffman))
			throw new ClassCastException("Tipo passato non valido"); 
		return new String(data + "|" + String.valueOf(frequency)); 
	}

}

File: KalkHead/huffman.cpp
#include "huffman.h"

Huffman::Huffman(const Huffman* x){
    frequency = x->frequency;
    data = x->data;
}

bool Huffman::operator==(const Tipo* t) const{
    const Huffman* x = dynamic_cast<const Huffman*>(t);

    if(!x)
        return false;

    if(frequency == x->frequency && data == x->data)
        return true;

    else
        return false;
}

bool Huffman::checkKey(const Tipo* x) const{
    const Huffman* p = dynamic_cast<const Huffman*>(x);
    if(p && p->data == data)
        return  true;
    else
        return false;
}

Huffman* Huffman::copia() const{
    return new Huffman(frequency, data);
}

Huffman* Huffman::operator+(const Tipo* x) const{
    const Huffman* t = dynamic_cast<const Huffman*>(x);

    if(t)
        return new Huffman(frequency + t->frequency, data + t->data);
    else
        return nullptr;

}

Huffman* Huffman::operator-(const Tipo*) const{
    throw new OperationNotAllowed();
}

Huffman* Huffman::operator *(const Tipo*) const{
    throw new OperationNotAllowed();
}

Huffman* Huffman::operator /(const Tipo*) const{
    throw new OperationNotAllowed();
}

Huffman& Huffman::operator =(const Tipo& t){
    const Huffman& p = dynamic_cast<const Huffman&>(t);
    if(this != &p){
        // Distruzione profonda dei campi
        frequency = p.frequency;
        data = p.data;
    }
    return *this;
}

bool Huffman::operator <(const Tipo* t) const{
    const Huffman* p = dynamic_cast<const Huffman*>(t);
    if(t && p){
        if(frequency < p->frequency)
            return true;
    }
    return false;
}

void Huffman::print(std::ostream &os) const{
    os << to_string();
}

std::ostream& operator<<(std::ostream& os, const Huffman* t){
    if(t!=nullptr){
        t->print(os);
    }
    return os;
}

std::string Huffman::getData() const{
    return data;
}

int Huffman::getFrequency() const{
    return this->frequency;
}

std::string Huffman::to_string() const{
    return data + "|" + std::to_string(frequency);
}

bool Huffman::operator>(const Tipo* t) const{
    const Huffman* p = dynamic_cast<const Huffman*>(t);
    if(t && p){
        if(frequency > p->frequency)
            return true;
    }
    return false;
}
bool Huffman::operator<=(const Tipo* t) const{
    const Huffman* p = dynamic_cast<const Huffman*>(t);
    if(t && p){
        if(frequency <= p->frequency)
            return true;
    }
    return false;
}
bool Huffman::operator>=(const Tipo* t) const{
    const Huffman* p = dynamic_cast<const Huffman*>(t);
    if(t){
        if(frequency >= p->frequency)
            return true;
    }
    return false;
}
